app/mempool/lane.go

```diff
@@ -0,0 +1,279 @@
+	if err != nil {
+		return err
+	}
+
+	l.mtx.Lock()
+	defer l.mtx.Unlock()
+
+	if err = l.laneMempool.Insert(ctx, tx); err != nil {
+		return err
+	}
+
+	l.txIndex[txInfo.Hash] = struct{}{}
+	return nil
+}
+
+// CountTx returns the total number of transactions in the lane's mempool.
+func (l *Lane) CountTx() int {
+	return l.laneMempool.CountTx()
+}
+
+// Remove removes a transaction from the lane's mempool.
+func (l *Lane) Remove(tx sdk.Tx) error {
+	txInfo, err := l.GetTxInfo(tx)
+	if err != nil {
+		return err
+	}
+
+	l.mtx.Lock()
+	defer l.mtx.Unlock()
+
+	if err = l.laneMempool.Remove(tx); err != nil {
+		return err
+	}
+
+	delete(l.txIndex, txInfo.Hash)
+	return nil
+}
+
+// Contains returns true if the lane's mempool contains the transaction.
+func (l *Lane) Contains(tx sdk.Tx) bool {
+	txInfo, err := l.GetTxInfo(tx)
+	if err != nil {
+		return false
+	}
+
+	l.mtx.RLock()
+	defer l.mtx.RUnlock()
+
+	_, exists := l.txIndex[txInfo.Hash]
+	return exists
+}
+
+// Match returns true if the transaction belongs to the lane.
+func (l *Lane) Match(ctx sdk.Context, tx sdk.Tx) bool {
+	return l.matchFn(ctx, tx)
+}
+
+// FillProposal fills the proposal with transactions from the lane mempool with the its own limit.
+// It returns the total size and gas of the transactions added to the proposal.
+// It also returns an iterator to the next transaction in the mempool and a list
+// of transactions that were removed from the lane mempool.
+func (l *Lane) FillProposal(
+	ctx sdk.Context,
+	proposal *Proposal,
+) (blockUsed BlockSpace, iterator sdkmempool.Iterator, txsToRemove []sdk.Tx) {
+	var (
+		transactionLimit BlockSpace
+		laneLimit        BlockSpace
+	)
+	// Get the transaction and lane limit for the lane.
+	transactionLimit = proposal.GetLimit(l.maxTransactionSpace)
+	laneLimit = proposal.GetLimit(l.maxLaneSpace)
+
+	// Select all transactions in the mempool that are valid and not already in the
+	// partial proposal.
+	for iterator = l.laneMempool.Select(ctx, nil); iterator != nil; iterator = iterator.Next() {
+		// If the total size used or total gas used exceeds the limit, we break and do not attempt to include more txs.
+		// We can tolerate a few bytes/gas over the limit, since we limit the size of each transaction.
+		if laneLimit.IsReachedBy(blockUsed) {
+			break
+		}
+
+		tx := iterator.Tx()
+		txInfo, err := l.GetTxInfo(tx)
+		if err != nil {
+			l.logger.Info("failed to get hash of tx", "err", err)
+
+			txsToRemove = append(txsToRemove, tx)
+			continue
+		}
+
+		// if the transaction is exceed the limit, we remove it from the lane mempool.
+		if transactionLimit.IsExceededBy(txInfo.BlockSpace) {
+			l.logger.Info(
+				"failed to select tx for lane; tx exceeds limit",
+				"tx_hash", txInfo.Hash,
+				"lane", l.name,
+			)
+
+			txsToRemove = append(txsToRemove, tx)
+			continue
+		}
+
+		// Add the transaction to the proposal.
+		// TODO: check if the transaction cannot be added here, it should also cannot be added afterward.
```

Comment:
are we doing this in this pr? if so just bumping

Reply:
deleted